sound/soc/sof/intel/cnl.c

```diff
@@ -171,23 +171,52 @@ static bool cnl_compact_ipc_compress(struct snd_sof_ipc_msg *msg,
 static int cnl_ipc_send_msg(struct snd_sof_dev *sdev,
 			    struct snd_sof_ipc_msg *msg)
 {
+	struct sof_intel_hda_dev *hdev = sdev->pdata->hw_pdata;
+	struct sof_ipc_cmd_hdr *hdr;
+	const struct sof_dsp_power_state target_state = {
+		.state = SOF_DSP_D0,
+		.substate = SOF_HDA_DSP_D0I0,
+	};
 	u32 dr = 0;
 	u32 dd = 0;
+	int ret;
 
 	if (cnl_compact_ipc_compress(msg, &dr, &dd)) {
 		/* send the message via IPC registers */
 		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDD,
 				  dd);
 		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
 				  CNL_DSP_REG_HIPCIDR_BUSY | dr);
-	} else {
-		/* send the message via mailbox */
-		sof_mailbox_write(sdev, sdev->host_box.offset, msg->msg_data,
-				  msg->msg_size);
-		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
-				  CNL_DSP_REG_HIPCIDR_BUSY);
+		return 0;
```

Comment:
you should clarify if we have a case where sending an ipc over the registers will result in the DSP exiting D0i3, in which case we could be no longer synchronized since the power state is not changed until the next IPC.

Reply:
The only case we have is a PM_GATE IPC over the registers to actually do the D0i3->D0i0 transition and this is always performed using the set_power_state() op that also synchronizes the state.